### ISSUE TITLE
Cannot run multiple setup types in the same repository

### ISSUE THREAD

**Issue by Fishbowler** [IcedFrisby/IcedFrisby#93]
We've got warnings in the repo for calling globalSetup() more than one, and that it "may cause unstable behavior".

I've run into that behaviour.

I've got multiple specs, each one requires one of two different setups (with/without JSON).
The before() of each spec runs the setup.
When run individually, each spec succeeds.
When run together with a mocha globbing selector, some specs don't run under the correct setup.

I've looked at the code and can't see an easy way to resolve this (e.g. some complex it-level config cloning). This feels like it could be a supported feature. I've got microservice API checks all bundled together rather than creating 2x the repos.

I couldn't write a node wrapper program that exec'd each spec in isolation, but is that dirtier than the problem?

**Comment by Fishbowler** [comment]
There are few other workarounds my team considered yesterday, but none of them are nice.

The best we've got so far is to use 1 globalSetup for retry/timeout/followRedirect, then use a plugin (or prototype?) to use in place of get/post that would set the json and baseUrl config for an individual request, done per-file in some contextual way. Not quite worked out the fine detail. This feels less work that solving it "for real" - I can't begin to get my head round how that could be done.

**Comment by paulmelnikow** [comment]
I'm inclined to get rid of the global state entirely. It's difficult to reason about and makes testing hard.

As an alternative, we could make a function that creates a partially-configured IcedFrisby factory, possibly using (internally) same constructor factory pattern we're using for plugins. Then a developer could create a short module exporting just the new Frisby constructor, import that in place of IcedFrisby, and get the same effect.

What do you think?

**Comment by Fishbowler** [comment]
Perfect - that sounds kinda similar to what I was gonna do locally for my repo of tests.

This definitely feels like a breaking change.

What if rather than having a global config, we had a global config registry?

```
frisby.registerConfig('config1',{
	request: {
		followRedirect: false,
		json: true
	},
	timeout: 5000,
	retry: 2
})

frisby.registerConfig('config2',{
	request: {
		json: false
	},
	timeout: 1000,
	retry: 0
})

frisby.createWithConfig('Simple Test', 'config1')
	.get('http://httpbin.org/get?foo=bar')
	...etc.

```
We could even still have the concept of a "default config" to use with `.create()`, but still have each test made with create() maintain its own clone/copy of the config.

**Comment by paulmelnikow** [comment]
Is there an advantage of having global state within IcedFrisby, over keeping that within the developer's module? Pure modules are my jam.

I suppose it lets you just import `icedfrisby` instead of `./my-icedfrisby`. Is it worth it?

**Comment by Fishbowler** [comment]
True. I already do this anyway so that I don't have to have the mixin at the top of every file.
So long as we do a top-notch job of example implementation, this would do lovely. Hmmmm. icedfrisby-examples repo maybe? Would be well up for implementing some of that.

**Comment by paulmelnikow** [comment]
Awesome! Yea, this seems essential enough that it probably belongs in our API docs, or if it seems out of place there, maybe a separate guide doc.

I wonder, would it be possible to handle global setup by overriding the static `create()` method, to invoke the superclass' `create()` method, and then run a few methods on the chain?

**Comment by Fishbowler** [comment]
That's a good shout. I could still do that on a file-by-file basis too, since each area of our APIs had a config file to store endpoint paths and the like. Nice idea!

Now to the real work though: unpicking global state.

**Comment by paulmelnikow** [comment]
It occurs to me… you could also implement the config registry as a plugin! Though the file-by-file solution might be cleaner.

By the way, would you be interested in collaborating on the plugin maintenance too? I'd be happy to move mine into the organization.

**Comment by paulmelnikow** [comment]
Ha, right, I remember, we had decided to try the monorepo thing first. 🤔

**Comment by Fishbowler** [comment]
Should be resolved by #106
